src/libraries/System.Management/src/System/Management/InteropClasses/WMIInterop.cs

```diff
@@ -490,12 +490,12 @@ public int GetNativeDataSize()
             return -1; // not a value type, so use -1
         }
 
-        public IntPtr MarshalManagedToNative(object obj)
+        public IntPtr MarshalManagedToNative(object? obj)
         {
-            return (IntPtr)obj;
+            return obj == null ? IntPtr.Zero : (IntPtr)obj;
```

Comment:
This one raises some red flags.  It suggests null actually hasn't been allowed, as this would have blown up if null was passed, and our stance with nullable annotations with regards to inputs on existing interface methods is that we should generally assume the least common denominator of known implementations, which here would mean non-nullable.

This could of course be classified as a bug, I'm just nervous in such situations it's representative of a larger class of issues.  Convince me I shouldn't be nervous :)

Reply:
Turns Out™ that this particular file doesn't have nullability enabled in the first place, so the change wasn't even necessary.  This change has been reverted.

Reply:
My question persists, though :) It's an implementation of the interface that doesn't accept nulls, yet we're saying all implementations of the interface allow nulls.  Just searching for ICustomMarshaler on github and clicking on a few of the first results, I see:
https://github.com/bitzhuwei/Kitware.VTK/blob/ce81a3cec7516488411fd767d18136a6034f08d6/Kitware.VTK/Kitware.VTK/vtkObjectMarshaler.cs#L26
https://github.com/thatistest/Ngit/blob/8b82d6bb58ae0440907648eef955849aafdb87e1/src/ShellNgitExten/IconMarshaler.cs#L40-L43
https://github.com/Melov/NetworkADB/blob/d86e2f0811ccbe677c12ab1078ae95ba789aba72/WindowsShell/Interop/ItemIdListMarshaler.cs#L38-L41
https://github.com/Melov/NetworkADB/blob/d86e2f0811ccbe677c12ab1078ae95ba789aba72/WindowsShell/Interop/ItemIdMarshaler.cs#L38-L41
https://github.com/jpark730/Fwankie/blob/589f3dfb004fbb91b056d34dac46d2989a05c267/3rdParty/MyGUI/Wrappers/MyGUI.Sharp/Core/InterfaceMarshaler.cs#L28
All of those fail if null is passed in.  That doesn't bode well for signaling to a caller of this interface that null is allowed.

Reply:
To which I refer to: https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/docs/coding-guidelines/api-guidelines/nullability.md#virtualabstract-methods-and-interfaces

> How common is it that an existing override written before the guarantee was put in effect would return `null`?

Multiple `ICustomMarshaler` implementations don't return `null` or won't accept `null`.  Many *do*.  I don't know if we can say *definitively* which is more common, but supporting `null` isn't *uncommon*.

> How widespread are overrides of the method in question? This contributes to (1).

There are fewer; if I use [GitHub as my guide](https://github.com/search?q=language%3Acsharp+override+marshalnativetomanaged&type=code), searching for C# code which contains *both* `override` *and* `MarshalNativeToManaged`, there's *only* 608 results (compare with 3,545 in the original query), and only one result on the first page is actually relevant, and it throws an exception:

https://github.com/qwer640/5678/blob/ddb85384030e15c4d35dfb6ba29f8342cf32795d/ksp_plugin_adapter/utf8_marshaler.cs#L58-L60

There are more interesting results later:

  * https://github.com/andrewkirillov/AForge.NET/blob/da91a56e2689d867849a3ba678f95a0701ed18ba/Sources/Video.DirectShow/Internals/CustomMarshalers.cs#L86
    `null` can be returned in the `override`.
  * https://github.com/NickeManarin/ScreenToGif/blob/7b71b39646a60661bc331ce490d054c298bb0fb2/ScreenToGif/Webcam/DirectShow/Marshaller.cs#L102
    `null` can be returned in the `override`.
  * https://github.com/mockingbirdnest/Principia/blob/50663002197a1909991de0f01cd829d88602317e/ksp_plugin_adapter/disposable_marshaler.cs#L23
    Always allocates a new value from `MarshalNativeToManaged()`; appears to support `MarshalManagedToNative(null)`.
  * https://github.com/ImpromptuNinjas/UltralightSharp/blob/c4582b46a85d12e4f18c1425d57fdc61e6bb6710/UltralightSharp/Utility/RefOnlyPtrWrapperMarshaller.cs#L7
    Overrides, but always returns a non-null value.

There aren't *no* overrides, and it's likewise a mixed bag: some will return `null`, some won't.

> How common is it to invoke the method via the base vs via a derived type that may narrow the return type to `T` from `T?`?

I'm not sure how to adequately answer this question in the context of `ICustomMarshaler`, as it's usually placed on return types or parameter types, and "marshaling magic happens".  The "direct caller" is the .NET runtime marshaler itself…and unit tests.

If we stick with the "marshaling magic happens" version, then it's very *uncommon* to invoke these methods *at all*.  I don't think this question applies?

> How common is it in the case of (3) for such invocations to then dereference the result rather than passing it off to something else that accepts a` T?`?

As with (3) and "magic happens", I submit that it's *very rare* for `ICustomMarshaler` method invocations to ever *directly* dereference the result.  Instead, it would be *implicit*:

```csharp
[return: MarshalAs (UnmanagedType.CustomMarshaler, MarshalTypeRef=typeof(Whatever))]
static extern SomeReturnType SomeMethod();
```

If `Whatever.MarshalNativeToManaged()` returns null, then `SomeMethod()` will return `null`, regardless of nullability annotations.  There isn't a good way (that I know of) to "flow" the `SomeReturnType` nullability semantics to the `Whatever.MarshalNativeToManaged()` nullability semantics (or vice versa).  (There likewise isn't a good way to statically ensure that `Whatever.MarshalNativeToManaged()` will even return a value of `SomeReturnType`; it could be *anything*…and result in a cast exception at runtime.)

If `Object.ToString()` returns `string?` because some (supposedly?) small-ish set code returned `null`, but it was "fine" because most `Object.ToString()` invocations aren't "chained" and instead are sent un-modified to e.g. `string.Format()`, how is `ICustomMarshaler` any different?  A decent set of code supports and uses `null`, and most such usage is "buried" inside runtime marshaling code and behind custom attributes.

The *callers* of `ICustomMarshaler` are few, but there are *lots* of *implementors*, and I'm not entirely keen on the idea of making, say, half of them need to disable nullability checks or alter their existing code in order to adopt nullability.

Reply:
>The callers of ICustomMarshaler are few, 

The Interop system really should be the only callers in practice and these are consumed from within the IL Stubs so nullability checks have no impact. Are there any examples of non-Interop callers? Unit tests? etc.

Reply:
> Multiple ICustomMarshaler implementations don't return null or won't accept null. Many do.

The thing is, it doesn't matter as much that many do: it matters more that some don't.  The whole point of nullable annotations is to issue warnings when a caller is doing something that might result in unexpected failure to due nulls being passed as arguments when the callee didn't allow them, or when the caller assumes null won't be returned but the callee may return null.  If our goal was to avoid warnings, we could just disable nullable annotations altogether, problem solved :)

If I have an interface that defines an argument as nullable, callers of that method can expect they can pass in null and it'll be fine, but if implementers of that method then don't accept null, either they've violated the contract, or the contract was wrong.  It's hard in cases like this one where the contract wasn't clear and we then have to infer what it was meant to be and the harm that may come if we get it wrong.  When in doubt, we typically err on the side of providing the most restrictive for arguments (non-nullable) and most permissive for return values (nullable) because that more accurately represents to callers the contraints they need to work under, that an implementation may not accept nulls, so be careful, and an implementation may return null, so be careful.  We do that because _typically_ there are orders of magnitude more callers of a given override / interface implementation than there are implementers of it.  It's the reason nullable annotations exist, to help flag for a dev when something unexpected might happen.

> The "direct caller" is the .NET runtime marshaler itself

This changes things, in that the claim is basically saying "we control all the callers that matter", yes?

Do we ever pass in null?  Are we totally happy with null returns?

Reply:
None of our call outs pass null arguments. The returns will require some additional investigation.

NativeToManaged:
https://github.com/dotnet/runtime/blob/cf842f10d5fc2f26cd973adfe338c9835caa1b8c/src/coreclr/vm/custommarshalerinfo.cpp#L172-L184

ManagedToNative:
https://github.com/dotnet/runtime/blob/cf842f10d5fc2f26cd973adfe338c9835caa1b8c/src/coreclr/vm/custommarshalerinfo.cpp#L204-L217

CleanUpNative:
https://github.com/dotnet/runtime/blob/cf842f10d5fc2f26cd973adfe338c9835caa1b8c/src/coreclr/vm/custommarshalerinfo.cpp#L237-L249

CleanUpManaged:
https://github.com/dotnet/runtime/blob/cf842f10d5fc2f26cd973adfe338c9835caa1b8c/src/coreclr/vm/custommarshalerinfo.cpp#L265-L276

Reply:
> None of our call outs pass null arguments

Then it begs the question, if these are the only callers, why do implementations need to special-case null?

Reply:
@stephentoub Unsure. It is quite possible non passing null was a later optimization. At this point it doesn't look like marking the argument as nullable is needed, but given what I am seeing returning null is okay. I am leaning more toward this change simply creates more questions than answers. If there are many implementers of the interface this could have a big impact on that group. Still questioning if the impact is worth the change for this particular interface.

Reply:
One other thing to follow up here would be to see how Mono consumes this interface.

/cc @lambdageek

Reply:
The Mono side is: https://github.com/dotnet/runtime/blob/2bdee5b5239facadad0ed7191ddc7143fe0d6a45/src/mono/mono/metadata/marshal-ilgen.c#L4685-L4715

plus surrounding code.

Mono thus appears to follow .NET and "null check" parameters and return values, in that eg. `ICustomMarshaler.MarshalNativeToManaged()` will *not* be invoked with a value of `IntPtr.Zero`, nor will `ICustomMarshaler.MarshalManagedToNative()` be invoked with a value of `null`.

Consequently, this change is not only unnecessary, but contrary to what the default behavior currently is.